tools/custom-stats-qc/1.0.0/custom-stats-qc__1.0.0.cwl

```diff
@@ -0,0 +1,154 @@
+cwlVersion: v1.1
+class: CommandLineTool
+
+# Extensions
+$namespaces:
+    s: https://schema.org/
+    ilmn-tes: http://platform.illumina.com/rdf/ica/
+$schemas:
+  - https://schema.org/version/latest/schemaorg-current-http.rdf
+
+# Metadata
+s:author:
+    class: s:Person
+    s:name: Sehrish Kanwal
+    s:email: sehrish.kanwal@umccr.org
+
+# ID/Docs
+id: custom-stats-qc--1.0.0
+label: custom-stats-qc v(1.0.0)
+doc: |
+    A tool to extract custom QC metrics from samtools stats output and convert to json format.
+
+# ILMN Resources Guide: https://support-docs.illumina.com/SW/ICA/Content/SW/ICA/RequestResources.htm
+hints:
+    ResourceRequirement:
+        ilmn-tes:resources:
+            tier: standard
+            type: standard
+            size: small
+        coresMin: 1
+        ramMin: 4000
+    DockerRequirement:
+        dockerPull: quay.io/umccr/alpine-jq:1.6
+
+requirements:
+  InlineJavascriptRequirement: {}
+  InitialWorkDirRequirement:
+    listing:
+      - entryname: "run-custom-qc.sh"
+        entry: |
+          #!/usr/bin/env sh
+
+          # Set up for failure
+          set -xeuo pipefail
+
+          # Sample metadata
+          JSON_STRING=\$( jq -n \\
+                            --arg sm "$(inputs.sample_id)" \\
+                            '{sampleID: $sm}' )
+          echo "$(inputs.output_json_filename).json"
+          echo \$JSON_STRING >> "$(inputs.output_json_filename).json"
+
+          # Extract average base quality
+          avg_base_qual=\$(grep -h "SN	average quality" "$(inputs.output_samtools_stats.path)" | cut -f 3)
+          JSON_STRING=\$( jq -n \\
+                            --arg des "average base quality" \\
+                            --arg src "SN	average quality" \\
+                            --arg val "$avg_base_qual" \\
+                            '{average_base_quality: {description: $des, source: $src, value: $val}}' )
+          echo \$JSON_STRING >> "$(inputs.output_json_filename).json"
+
+          # Extract yield_mapped_bp
+          yield_mapped_bp=\$(grep -h "SN	bases mapped (cigar)" "$(inputs.output_samtools_stats.path)" | cut -f 3)
+          JSON_STRING=\$( jq -n \\
+                            --arg des "yield mapped bp" \\
+                            --arg src "SN	bases mapped (cigar)" \\
+                            --arg val "$yield_mapped_bp" \\
+                            '{yield_mapped_bp: {description: $des, source: $src, value: $val}}' )
+          echo \$JSON_STRING >> "$(inputs.output_json_filename).json"
```

Comment:
A lot of WET typing here.  

What if we had a function that took in four arguments (pattern, stats file, key, description) and returned the string? 

A function like so should do:

```
get_json_str_from_stats_file(){
  local pattern="$1"
  local stats_file="$2"
  local key="$3"
  local description="$4"
  local value
  # Get value from file and then parse through jq
  value="$(grep "${pattern}" "${stats_file}" | cut -f2)"
  # Return json string
  jq \
    --null-input \
    --arg "des" "${description}" \
    --arg "src" "${pattern}" \
    --arg "val" "${value}" \
    --arg "key_name" "${key_name}" \
    '{$key_name: {description: $des, source: $src, value: ${val}}'
}
```

We could then call the function like so
```
yield_mapped_bp_json="$(get_json_str_from_stats_file "bases mapped (cigar)" "${summary_stats_file}" "yield_mapped_bp")"
```

We can then just run the following line at the end
```
echo "${avg_base_qual_json} ${yield_mapped_bp_json} ..." | jq '[inputs] | add' > output_file_name
```

Reply:
Alternatively, just use the one variable

```
JSON_STRING="${JSON_STRING}"\n"$(get_json_str_from_stats_file "bases mapped (cigar)" "${summary_stats_file}" "yield_mapped_bp")"
```

Reply:
Or just do this:

```
get_json_str_from_stats_file "bases mapped (cigar)" "${summary_stats_file}" "yield_mapped_bp" >> TMP_OUTPUT_FILE.json
```

Reply:
Finally went with a function that creates the json via jq https://github.com/umccr/cwl-ica/blob/52ee1dd285bbb8d113e4e53f9dbf23c38cb9c750/tools/custom-stats-qc/1.0.1/custom-stats-qc__1.0.1.cwl#L46

And used a single variable `JSON_STRING` to store the jq output e.g. https://github.com/umccr/cwl-ica/blob/52ee1dd285bbb8d113e4e53f9dbf23c38cb9c750/tools/custom-stats-qc/1.0.1/custom-stats-qc__1.0.1.cwl#L77

Couple metrics (https://github.com/umccr/cwl-ica/blob/52ee1dd285bbb8d113e4e53f9dbf23c38cb9c750/tools/custom-stats-qc/1.0.1/custom-stats-qc__1.0.1.cwl#L92 and https://github.com/umccr/cwl-ica/blob/52ee1dd285bbb8d113e4e53f9dbf23c38cb9c750/tools/custom-stats-qc/1.0.1/custom-stats-qc__1.0.1.cwl#L108) required a bit more custom calculations which required some WET typing.